### PR TITLE
Incremental view maintenance

### PR DESCRIPTION
**Pull request by yjhjstz**
closes: #ISSUE_Number

---

### Change logs

background: 
* https://github.com/orgs/cloudberrydb/discussions/36
* https://wiki.postgresql.org/wiki/Incremental_View_Maintenance
* https://www.postgresql.org/message-id/flat/20230601235909.0e1572c27e59112f9d0cbe86%40sraoss.co.jp

### Why are the changes needed?

Base on patch v28 from Postgres, here is design code: https://hashdata.feishu.cn/docx/RdS6dND7xotUhLxQi1UctzCVnih

The challenge lies in how to implement IVM in a distributed environment.

### Does this PR introduce any user-facing change?

_If yes, please clarify the previous behavior and the change this PR proposes._

### How was this patch tested?

Add extra test to regress:
- ivm.sql
- incremental_matview.sql

### Contributor's Checklist

Here are some reminders and checklists before/when submitting your pull request, please check them:

- [ ] Make sure your Pull Request has a clear title and commit message. You can take [git-commit](https://github.com/cloudberrydb/cloudberrydb/blob/main/.gitmessage) template as a reference.
- [x] Sign the Contributor License Agreement as prompted for your first-time contribution(*One-time setup*).
- [ ] Learn the [coding contribution guide](https://cloudberrydb.org/contribute/code), including our code conventions, workflow and more.
- [ ] List your communication in the [GitHub Issues](https://github.com/cloudberrydb/cloudberrydb/issues) or [Discussions](https://github.com/orgs/cloudberrydb/discussions) (if has or needed).
- [x] Document changes.
- [x] Add tests for the change
- [x] Pass `make installcheck`
- [x] Pass `make -C src/test installcheck-cbdb-parallel`
- [ ] Feel free to @cloudberrydb/dev team for review and approval when your PR is ready🥳

### CODE REVIEW

File: src/backend/commands/createas.c [hunk 1]
```diff
@@ -363,6 +393,22 @@ ExecCreateTableAs(ParseState *pstate, CreateTableAsStmt *stmt,
...
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("mutable function is not supported on incrementally maintainable materialized view"),
+					 errhint("functions must be marked IMMUTABLE")));
+
+		check_ivm_restriction((Node *) query);
+
+		/* For IMMV, we need to rewrite matview query */
+		query = rewriteQueryForIMMV(query, into->colNames);
+		query_immv = copyObject(query);
+	}
```
Comment (avamingli) [inline src/backend/commands/createas.c:410]: New variable `query_immv` make us lost some logic like `rewritten = QueryRewrite(query);` as below which IMV patch28 didn't do that.

File: src/backend/commands/createas.c [hunk 2]
```diff
@@ -802,3 +1031,778 @@ GetIntoRelOid(QueryDesc *queryDesc)
...
+							(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							 errmsg("ORDER BY clause is not supported on incrementally maintainable materialized view")));
+				if (qry->limitOffset != NULL || qry->limitCount != NULL)
+					ereport(ERROR,
+							(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							 errmsg("LIMIT/OFFSET clause is not supported on incrementally maintainable materialized view")));
+				// if (qry->distinctClause)
+				// 	ereport(ERROR,
+				// 			(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				// 			 errmsg("DISTINCT is not supported on incrementally maintainable materialized view")));
```
[the remark this answers is not in the document]
Comment (yjhjstz) [reply src/backend/commands/createas.c]: clean up.

File: src/backend/commands/createas.c [hunk 3]
```diff
@@ -802,3 +1031,778 @@ GetIntoRelOid(QueryDesc *queryDesc)
...
+	 * The type of lock should be determined here, because if we check the
+	 * view definition at maintenance time, we need to acquire a weaker lock,
+	 * and upgrading the lock level after this increases probability of
+	 * deadlock.
+	 */
+
+	// rte = list_nth(qry->rtable, 0);
+	// if (list_length(qry->rtable) > 1 || rte->rtekind != RTE_RELATION)
+	// 	ex_lock = true;
+	ex_lock = true;
```
[the remark this answers is not in the document]
Comment (yjhjstz) [reply src/backend/commands/createas.c]: clean up.

File: src/backend/commands/createas.c [hunk 4]
```diff
@@ -802,3 +1031,778 @@ GetIntoRelOid(QueryDesc *queryDesc)
...
+ * CreateIndexOnIMMV
+ *
+ * Create a unique index on incremental maintainable materialized view.
+ * If the view definition query has a GROUP BY clause, the index is created
+ * on the columns of GROUP BY expressions. Otherwise, if the view contains
+ * all primary key attritubes of its base tables in the target list, the index
+ * is created on these attritubes. In other cases, no index is created.
+ */
+void
+CreateIndexOnIMMV(Query *query, Relation matviewRel)
```
[the remark this answers is not in the document]
Comment (yjhjstz) [reply src/backend/commands/createas.c:1523]: index is not prerequisite, just accelerate. Group by is supported without index.

File: src/backend/commands/createas.c (hunk 1 above)
Comment (yjhjstz) [reply src/backend/commands/createas.c:410]: query_immv is just used to create triggers for materialized view.

File: src/backend/commands/createas.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (avamingli) [reply src/backend/commands/createas.c:1523]: Why don't  create index to accelerate?

File: src/backend/commands/createas.c (hunk 1 above)
Comment (avamingli) [reply src/backend/commands/createas.c:410]: Why avoid rewrite?

File: src/backend/commands/createas.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (yjhjstz) [reply src/backend/commands/createas.c:1523]: index maintenance in transaction also cost time, now insert is much slow with ivm.

File: src/backend/commands/createas.c (hunk 1 above)
Comment (yjhjstz) [reply src/backend/commands/createas.c:410]: we just need base tables , not include rewrite table from rule.

**Comment by avamingli** [comment]
Hi,  I want to see what we changed besides pg's patch, but it's hard to tell as there are separate commits of our own codes. 
Could we squash them to one for easy to review? And we will do it finally when merging the pr.
And some commits should be kept as pg's, ex: [Add documentations about Incremental View Maintenance]

**Comment by yjhjstz** [comment]
> Hi, I want to see what we changed besides pg's patch, but it's hard to tell as there are separate commits of our own codes.
> Could we squash them to one for easy to review? And we will do it finally when merging the pr.
> And some commits should be kept as pg's, ex: [Add documentations about Incremental View Maintenance]

image

tried to arrange [Add documentations about Incremental View Maintenance].

**Pull request merged**